File: sdk/typescript/src/crypto.ts
/**
 * Cryptographic signing for ADO binding (SPEC §4.6)
 *
 * Uses the Node.js built-in `crypto` module (no external dependencies).
 * Ed25519 support requires Node.js >= 15 with SubtleCrypto.
 */

import { createHash, createHmac, timingSafeEqual } from "node:crypto";
import type { DelegationRules } from "./types.js";

/**
 * SHA-256 hash of canonical JSON (SPEC §4.6).
 */
export function sha256Hex(data: string): string {
  return createHash("sha256").update(data, "utf8").digest("hex");
}

/**
 * Builds the canonical signing payload for an ADO (SPEC §4.6).
 *
 * Covers: decision_id, authorized_dsal, authority, expires_at, proposal_hash,
 * nonce, delegation_rules.
 */
export function canonicalPayload(
  decisionId: string,
  authorizedDsal: number,
  authority: string,
  expiresAt: string,
  proposalHash: string,
  nonce: string,
  delegationRules: DelegationRules,
): string {
  return JSON.stringify({
    decision_id: decisionId,
    authorized_dsal: authorizedDsal,
    authority,
    expires_at: expiresAt,
    proposal_hash: proposalHash,
    nonce,
    delegation_rules: delegationRules,
  });
}

/**
 * HMAC-SHA256 signature over canonical payload (SPEC §4.6 minimum requirement).
 */
export function hmacSign(key: string, payload: string): string {
  return createHmac("sha256", key).update(payload, "utf8").digest("hex");
}

/**
 * Verify an HMAC-SHA256 signature using timing-safe comparison.
 */
export function hmacVerify(key: string, payload: string, expected: string): boolean {
  const computed = hmacSign(key, payload);
  const a = Buffer.from(computed, "hex");
  const b = Buffer.from(expected, "hex");
  if (a.length !== b.length) return false;
  return timingSafeEqual(a, b);
}

File: runtime/rust/src/crypto.rs
//! Cryptographic signing for ADO binding (SPEC §4.6)

use hmac::{Hmac, Mac};
use sha2::Sha256;
use serde_json::Value;

type HmacSha256 = Hmac<Sha256>;

/// Canonical payload for signing (SPEC §4.6)
///
/// Covers: decision_id, authorized_dsal, authority, expires_at, constraints,
/// proposal_hash, nonce, delegation_rules.
pub fn canonical_payload(
    decision_id: &str,
    authorized_dsal: u8,
    authority: &str,
    expires_at: &str,
    proposal_hash: &str,
    nonce: &str,
    delegation_rules: &Value,
) -> Vec<u8> {
    let payload = serde_json::json!({
        "decision_id": decision_id,
        "authorized_dsal": authorized_dsal,
        "authority": authority,
        "expires_at": expires_at,
        "proposal_hash": proposal_hash,
        "nonce": nonce,
        "delegation_rules": delegation_rules,
    });

    serde_json::to_string(&payload)
        .expect("canonical payload serialization must not fail")
        .into_bytes()
}

/// HMAC-SHA256 signature over canonical payload.
///
/// Minimum binding requirement from SPEC §4.6.
/// Use Ed25519 feature flag for production deployments.
pub fn hmac_sign(key: &[u8], payload: &[u8]) -> String {
    let mut mac = HmacSha256::new_from_slice(key)
        .expect("HMAC accepts any key size");
    mac.update(payload);
    hex::encode(mac.finalize().into_bytes())
}

/// Verify an HMAC-SHA256 signature.
pub fn hmac_verify(key: &[u8], payload: &[u8], expected_signature: &str) -> bool {
    let computed = hmac_sign(key, payload);
    // Constant-time comparison
    use sha2::Digest;
    let a = Sha256::digest(computed.as_bytes());
    let b = Sha256::digest(expected_signature.as_bytes());
    a == b
}
